src/decoder.rs

```diff
@@ -205,7 +204,7 @@ impl<R: Read> Decoder<R> {
                     let frame = self.frame.clone().unwrap();
                     let scan = parse_sos(&mut self.reader, &frame)?;
 
-                    if frame.coding_process == CodingProcess::DctProgressive && self.coefficients.is_empty() {
+                    if self.coefficients.is_empty() {
```

Comment:
Previously it would only need to keep one row of coefficients in memory at once. This makes it allocate coefficients for the whole image, resulting in a huge increase in peak memory usage.